mockgen/mockgen.go

```diff
@@ -223,6 +224,12 @@ func sanitize(s string) string {
 }
 
 func (g *generator) Generate(pkg *model.Package, pkgName string, outputPackagePath string) error {
+	// If the pkgName has _test as a suffix, then the package must be
+	// prepended.
+	if strings.HasSuffix(pkgName, "_test") {
```

Comment:
Are you sure this is the right test? For example, the input package could be the test package. Wouldn't the right test be that the input and output packages are not the same packages?

Reply:
That is what I thought originally, however it turns out to be a decent test as the case you described is already covered. For example if you look at the sample, it places the mocks in a new directory and the package is `mock_sample`. This all works.

The `test` package is a special case where it co-exists in the same directory, but has a different package name. Most of the mockgen code is looking at directories, and not package names.

Reply:
Sorry about that. I think I wrote this part and assumed that each dir only has one package. Must have forgotten about the _test exception.

I can't really tell from the code if this is going to always do the right thing. Would you mind adding a test case where the input and out are the same package ending in _test? That's the case where I'm concerned this might not do the right thing.

Reply:
I added a test and found after further investigation that I could do as you originally suggested. I've switched to this method.